enhancements/installer/coarse-grained-exit-codes.md

```diff
@@ -0,0 +1,159 @@
+
+## Summary
+
+`openshift-installer` callers will now get more granular exit codes.
+Callers of `openshift-installer` can expect non-zero exit codes *that may change over time* as
+the `openshift-installer` evolves over time.
+`openshift-installer` provides *no guarantee* that exit codes will not change as result of changes like
+greater granularity, lower granularity, re-organization, or other needs.
+`openshift-installer` *does guarantee* that a particular exit code will not be re-used to have a
+different meaning in the same y-level version.
+`openshift-installer` will make reasonable efforts to avoid re-using a particular exit code to have
+different meanings across y-level versions, but for sufficiently compelling reasons may do so.
+
+## Motivation
+
+Automated categorization of failures can assist in efforts like
+1. reliability determination of provision on various clouds
+2. directing effort to proper teams depending on failure modes
+3. tracking of which installation stages have problems on various platforms
+4. automated retry of infrastructure failures by CI step registry
+
+These are reasons why the data is useful and by providing it, other teams will be able to build these
+secondary system.
+Other teams *must properly handle previously unknown exit codes*.
+
+### Goals
+
+1. Coarse grained (order of 10 or less values), stage-based exit codes to assist in categorizing
+   failures, but not debugging failures.
+
+### Non-Goals
+
+1. Detailed, machine-parseable, cause-based failure information is outside the scope of this enhancement.
+
+## Proposal
+
+The installer knows what stage of the installation has failed when it exits.
+The stages include
+1. install-config handling
+2. infrastructure creation
+3. bootstrapping
+4. wait-for-cluster-install
+5. destroy
+6. everything else
+
+The installer will not introspect errors to provide deeper analysis, but these stages (and whatever future
+ones the install team identifies) may have unique exit codes when they fail.
+Categorizing failures in this way will allow machine categorization of the type of failure by other systems.
+The CI step registry is a good example of a consumer, but issuing a destroy after an infrastructure creation
+and retrying the `openshift-install` seems like a fairly common use-case since infrastructure creation
+failures often happen for environmental reasons and are reasonably likely to succeed on a retry.
+
+### Known Exit Codes
+
+Key things to remember:
+1. Callers of `openshift-installer` can expect non-zero exit codes *that may change over time* as
+   the `openshift-installer` evolves over time.
+2. `openshift-installer` provides *no guarantee* that exit codes will not change as result of changes like
+   greater granularity, lower granularity, re-organization, or other needs.
+3. `openshift-installer` *does guarantee* that a particular exit code will not be re-used to have a
+   different meaning in the same y-level version.
```

Comment:
How will a user know what the exit codes for a given version mean? Docs only? Sub-command help? Other?

Reply:
> How will a user know what the exit codes for a given version mean? Docs only? Sub-command help? Other?

Keeping this enhancement up to date seems reasonable.  I wouldn't get carried away with documenting what amount to best effort codes.

Reply:
Maybe human readable stderr output like:
`exit(1) -- Failed to parse install-config`